File: backend-dotnet/Application/Commands/Posts/CreatePostCommandHandler.cs
using Application.Common;
using Application.DTOs;
using Application.Interfaces;
using Domain.Entities;
using Domain.ValueObjects;
using Microsoft.Extensions.Logging;

namespace Application.Commands.Posts
{
    public class CreatePostCommandHandler : ICommandHandler<CreatePostCommand, PostListItemDto>
    {
        private readonly IPostRepository _postRepository;
        private readonly IUserRepository _userRepository;
        private readonly ITagRepository _tagRepository;
        private readonly IPostTagRepository _postTagRepository;
        private readonly IBlobService _blobService;
        private readonly IElasticService _elastic;
        private readonly ICacheService _cacheService;
        private readonly ILogger<CreatePostCommandHandler> _logger;

        public CreatePostCommandHandler(
            IPostRepository postRepository,
            IUserRepository userRepository,
            ITagRepository tagRepository,
            IPostTagRepository postTagRepository,
            IBlobService blobService,
            IElasticService elastic,
            ICacheService cacheService,
            ILogger<CreatePostCommandHandler> logger)
        {
            _postRepository = postRepository;
            _userRepository = userRepository;
            _tagRepository = tagRepository;
            _postTagRepository = postTagRepository;
            _blobService = blobService;
            _elastic = elastic;
            _cacheService = cacheService;
            _logger = logger;
        }

        public async Task<PostListItemDto> HandleAsync(CreatePostCommand command)
        {
            _logger.LogInformation("Creating post for user {UserId}", command.UserId);

            string? imageUrl = null;
            if (command.Dto.Image != null && command.Dto.Image.Length > 0)
            {
                imageUrl = await _blobService.UploadAsync(command.Dto.Image.OpenReadStream(), command.Dto.Image.FileName, BlobContainers.Post);
                _logger.LogInformation("Image uploaded successfully");
            }

            var post = new Post
            {
                Id = Guid.NewGuid(),
                Caption = command.Dto.Caption,
                ImageUrl = imageUrl,
                CreatedAt = DateTime.UtcNow,
                UserId = command.UserId,
                Visibility = command.Dto.Visibility
            };

            _postRepository.Add(post);

            // Handle hashtags
            await ProcessHashtagsAsync(post, command.Dto.Caption ?? string.Empty);

            await _postRepository.SaveChangesAsync();

            // Index for Elasticsearch
            await IndexPostAsync(post);

            // Clear relevant caches
            _cacheService.RemoveByPattern("public_posts");
            _cacheService.RemoveByPattern($"user_feed_{command.UserId}");

            _logger.LogInformation("Post {PostId} created successfully", post.Id);

            return await BuildPostListItemDto(post);
        }

        private async Task ProcessHashtagsAsync(Post post, string caption)
        {
            var postCaption = PostCaption.Create(caption);
            if (!postCaption.HasHashtags) return;

            var existingTags = await _tagRepository.GetExistingTagsAsync(postCaption.Hashtags);

            foreach (var tagName in postCaption.Hashtags)
            {
                if (!existingTags.TryGetValue(tagName, out var tag))
                {
                    tag = await _tagRepository.CreateTagAsync(tagName);
                    existingTags[tagName] = tag;
                }

                await _postTagRepository.CreatePostTagAsync(post.Id, tag.Id);

                await _elastic.IndexTagAsync(new TagIndex
                {
                    Id = tag.Id.ToString(),
                    Name = tag.Name
                });
            }
        }

        private async Task IndexPostAsync(Post post)
        {
            try
            {
                var user = await _userRepository.GetByIdAsync(post.UserId);
                if (user != null)
                {
                    await _elastic.IndexPostAsync(new PostIndex
                    {
                        Id = post.Id,
                        Caption = post.Caption ?? string.Empty,
                        CreatedAt = post.CreatedAt,
                        Username = user.Username
                    });
                }
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Failed to index post {PostId}", post.Id);
                // Don't throw - indexing failure shouldn't fail post creation
            }
        }

        private async Task<PostListItemDto> BuildPostListItemDto(Post post)
        {
            var user = await _userRepository.GetByIdAsync(post.UserId);
            
            return new PostListItemDto
            {
                Id = post.Id,
                Caption = post.Caption,
                ImageUrl = post.ImageUrl,
                CreatedAt = post.CreatedAt,
                Likes = 0, // New post has no likes
                Liked = false,
                User = user != null ? new SimpleUserDto
                {
                    Id = user.Id,
                    Username = user.Username,
                    ProfileImageUrl = user.ProfileImageUrl
                } : null!
            };
        }

    }
}

File: backend-java/src/main/java/com/jerrygram/application/commands/posts/CreatePostCommandHandler.java
package com.jerrygram.application.commands.posts;

import com.jerrygram.application.common.BlobContainers;
import com.jerrygram.application.common.HashtagExtractor;
import com.jerrygram.application.common.PostIndex;
import com.jerrygram.application.dtos.PostListItemDto;
import com.jerrygram.application.dtos.SimpleUserDto;
import com.jerrygram.application.interfaces.IBlobService;
import com.jerrygram.application.interfaces.ICacheService;
import com.jerrygram.application.interfaces.ICommandHandler;
import com.jerrygram.application.interfaces.IElasticService;
import com.jerrygram.domain.entities.Post;
import com.jerrygram.domain.entities.PostTag;
import com.jerrygram.domain.entities.Tag;
import com.jerrygram.infrastructure.repositories.PostRepository;
import com.jerrygram.infrastructure.repositories.PostTagRepository;
import com.jerrygram.infrastructure.repositories.TagRepository;
import com.jerrygram.infrastructure.repositories.UserRepository;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;

@Service
@RequiredArgsConstructor
@Slf4j
public class CreatePostCommandHandler implements ICommandHandler<CreatePostCommand, PostListItemDto> {

    private final PostRepository postRepository;
    private final UserRepository userRepository;
    private final PostTagRepository postTagRepository;
    private final TagRepository tagRepository;
    private final IElasticService elasticService;
    private final IBlobService blobService;
    private final ICacheService cacheService;

    @Override
    @Transactional
    public PostListItemDto handle(CreatePostCommand command) {
        var dto = command.getCreatePostDto();
        var authorId = command.getAuthorId();
        
        log.info("Creating post for user: {}", authorId);

        var author = userRepository.findById(authorId)
                .orElseThrow(() -> new IllegalArgumentException("User not found"));

        // Upload image if provided
        String imageUrl = null;
        if (dto.getImage() != null && !dto.getImage().isEmpty()) {
            try {
                imageUrl = blobService.upload(dto.getImage(), BlobContainers.POSTS);
                log.info("Image uploaded successfully: {}", imageUrl);
            } catch (Exception e) {
                log.error("Failed to upload image for post", e);
                throw new RuntimeException("Failed to upload image", e);
            }
        }

        var post = Post.builder()
                .caption(dto.getCaption())
                .imageUrl(imageUrl)
                .visibility(dto.getVisibility())
                .userId(authorId)
                .build();

        postRepository.save(post);
        
        // Extract and save hashtags using PostCaption value object
        var hashtags = post.getHashtags();
        for (String hashtag : hashtags) {
            // Find or create tag
            var tag = tagRepository.findByName(hashtag)
                    .orElseGet(() -> {
                        var newTag = Tag.builder()
                                .name(hashtag)
                                .build();
                        return tagRepository.save(newTag);
                    });
            
            // Create PostTag relationship
            var postTag = PostTag.builder()
                    .postId(post.getId())
                    .tagId(tag.getId())
                    .post(post)
                    .tag(tag)
                    .build();
            postTagRepository.save(postTag);
        }
        
        // Index in Elasticsearch
        try {
            elasticService.indexPost(PostIndex.builder()
                    .id(post.getId().toString())
                    .caption(post.getCaption())
                    .imageUrl(post.getImageUrl())
                    .authorId(author.getId().toString())
                    .authorUsername(author.getUsername())
                    .visibility(String.valueOf(post.getVisibility().ordinal()))
                    .likesCount(post.getLikesCount())
                    .commentsCount(post.getCommentsCount())
                    .tags(hashtags)
                    .createdAt(post.getCreatedAt())
                    .isActive(true)
                    .build());
        } catch (Exception e) {
            log.warn("Failed to index post in Elasticsearch: {}", post.getId(), e);
        }
        
        // Invalidate related caches
        invalidateRelatedCaches(post, hashtags);
        
        log.info("Post {} created successfully by user: {} with {} hashtags", 
                post.getId(), authorId, hashtags.size());

        return PostListItemDto.builder()
                .id(post.getId())
                .caption(post.getCaption())
                .imageUrl(post.getImageUrl())
                .createdAt(post.getCreatedAt())
                .likes(post.getLikesCount())
                .liked(false) // New post has no likes
                .user(SimpleUserDto.builder()
                        .id(author.getId())
                        .username(author.getUsername())
                        .profileImageUrl(author.getProfileImageUrl())
                        .build())
                .score(0.0)
                .build();
    }
    
    private void invalidateRelatedCaches(Post post, List<String> hashtags) {
        try {
            // Invalidate public posts cache
            cacheService.deleteByPattern("public_posts_page_*");
            
            // Invalidate user feed cache
            cacheService.delete("user_feed_" + post.getUserId());
            
            // Invalidate hashtag-related autocomplete caches
            for (String hashtag : hashtags) {
                cacheService.delete("autocomplete:#" + hashtag);
                cacheService.delete("autocomplete:" + hashtag);
                
                if (hashtag.length() >= 2) {
                    String prefix = hashtag.substring(0, 2);
                    cacheService.deleteByPattern("autocomplete:#" + prefix + "*");
                    cacheService.deleteByPattern("autocomplete:" + prefix + "*");
                }
            }
            
            log.debug("Cache invalidated for new post: {}", post.getId());
        } catch (Exception e) {
            log.warn("Failed to invalidate cache for post: {}", post.getId(), e);
        }
    }
}